### ISSUE TITLE
Unexpected error when using libc++ (call to deleted function 'from_chars')

### ISSUE THREAD

**Issue by aleck099** [llvm/llvm-project#59374]
### Code
<details>

Compile with `clang++ -std=c++17 -stdlib=libc++`

```cpp
#include <charconv>
#include <string>
#include <stdexcept>

template<typename T>
T decode_num(std::string_view s) {
    T r;
    auto ec = std::from_chars(s.data(), s.data() + s.size(), r);
    if (ec.ec == std::errc())
        return r;
    else
        throw std::runtime_error("decode_num");
}

int main() {
    const char fs[]{ "0.87" };
    float r = decode_num<float>(fs);
    return 0;
}
```
</details>

### Error:
<details>

```console
t1.cpp:8:15: error: call to deleted function 'from_chars'
    auto ec = std::from_chars(s.data(), s.data() + s.size(), r);
              ^~~~~~~~~~~~~~~
t1.cpp:17:15: note: in instantiation of function template specialization 'decode_num<float>' requested here
    float r = decode_num<float>(fs);
              ^
/usr/bin/../include/c++/v1/charconv:113:19: note: candidate function has been explicitly deleted
from_chars_result from_chars(const char*, const char*, bool, int = 10) = delete;
                  ^
/usr/bin/../include/c++/v1/charconv:588:1: note: candidate template ignored: requirement 'is_integral<float>::value' was not satisfied [with _Tp = float]
from_chars(const char* __first, const char* __last, _Tp& __value)
^
/usr/bin/../include/c++/v1/charconv:595:1: note: candidate function template not viable: requires 4 arguments, but 3 were provided
from_chars(const char* __first, const char* __last, _Tp& __value, int __base)
^
1 error generated.
```
</details>

### Versions:
Clang / libc++: 14.0.6

### Other information
Error does **NOT** occur when:
- compile with GCC or MSVC
- without `-stdlib=libc++`

**Comment by EugeneZelenko** [comment]
Could you please try 15 or `main` branch? https://godbolt.org should be helpful.

**Comment by philnik777** [comment]
`std::from_chars` isn't implemented yet. Please check https://libcxx.llvm.org/Status/Cxx17.html (or the status page where the corresponding paper is listed) before reporting unimplemented papers as bugs. For a fast check you can also look at the feature test macros (in this case `__cpp_lib_to_chars`, which isn't set).

**Comment by ceztko** [comment]
To whom it may concern: the paper partially implemented is [P0067r5](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0067r5.html). It proved to be a pain to implement to [MSVC ](https://www.youtube.com/watch?v=4P_kbF0EbZM) team (source [code](https://github.com/microsoft/STL/blob/main/stl/inc/charconv)) and possibly it was implemented in [gcc](https://gcc.gnu.org/pipermail/libstdc++/2020-July/050633.html) with an hack. Do you have an idea on how this could be delivered to llvm as well? I would definitely not blame the team if you follow the gcc path, after all only a fraction of the potential users would need the round trip feature of `to_chars`, `from_chars`. To all other people that just want to use `std::from_chars` I implemented a small incomplete [filler](https://github.com/pdfmm/pdfmm/blob/da0c368a564ad024d2c359b2a567f64f948913f4/src/pdfmm/private/charconv_compat.h) that uses [`fastfloat`](https://github.com/fastfloat/fast_float) behind the scenes.

**Comment by philnik777** [comment]
> To whom it may concern: the paper partially implemented is [P0067r5](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0067r5.html). It proved to be a pain to implement to [MSVC ](https://www.youtube.com/watch?v=4P_kbF0EbZM) team (source [code](https://github.com/microsoft/STL/blob/main/stl/inc/charconv)) and possibly it was implemented in [gcc](https://gcc.gnu.org/pipermail/libstdc++/2020-July/050633.html) with an hack. Do you have an idea on how/when this could be delivered to llvm as well? I would definitely not blame the team if you follow the gcc path, after all only a fraction of the potential users would need the round trip feature of `to_chars`, `from_chars`. To all other people that just want to use `std::from_chars` I implemented a small incomplete [filler](https://github.com/pdfmm/pdfmm/blob/da0c368a564ad024d2c359b2a567f64f948913f4/src/pdfmm/private/charconv_compat.h) that uses [`fastfloat`](https://github.com/fastfloat/fast_float) behind the scenes.

The fastest way to get it is to implement it and provide the implementation to libc++. Otherwise you have to hope that someone feels like implementing it. For the most part, libc++ is being worked on by volunteers. Only a relatively small part of the code is provided by any companies.

**Comment by ceztko** [comment]
I understand, thanks for pointing it out. Stephan T. Lavavej of MSVC team explained he worked about 1 year on this, which should be scaring for anyone with a job or any other activity. This also a reason why hacks or code "stealing" (respecting licenses, OFC) should be seriously considered to implement this paper. Sorry but I'm not devoted enough to llvm to be a contributor, especially on this subject: with my post I wanted to leave a note to anybody passing by explaining why `std::from_chars` is still not in libc++ and offer other viable options.

**Comment by mordante** [comment]
Even when using external code it's not a trivial task. Stephan contributed MSVC's version of `to_chars` to libc++. I've changed that code to libc++ code was quite a bit of work. (Obviously a lot less work than doing the entire work myself.)
When using external code there still are a lot of tests to be written. So yes this is a large project. I might take it up at some point, but it's low on my priority list. Not because it's not important, but just due to the amount of work.

**Comment by h-vetinari** [comment]
Stumbled over this while looking for the status of P0065R5; took me a while to unearth some old comments on this topic in https://reviews.llvm.org/D70631:

> @h-vetinari: I was under the [impression](https://twitter.com/lemire/status/1417215574657179657) that https://github.com/llvm/llvm-project/commit/87c016078ad72c46505461e4ff8bfa04819fe7ba did most of the work for from_chars?

> @mordante: Thanks for both links I wasn't aware of this effort!
At the moment I'm too busy with `<format>` to look at `from_chars`, but this is useful when somebody starts working on `from_chars`.
(I might look at it when I finished the larger projects I'm working on.)

I guess we can all but hope[^1] that the stack of C++20/23 `<format>` work will be done _eventually_. 😅 

[^1]: I consider it far outside my skillset to try to contribute myself here, unfortunately; even then, I'm still waiting for the move to GH, really not interested in working with Phab.

**Comment by mordante** [comment]
> I guess we can all but hope[1](#user-content-fn-1-70c01fc6cb56f31b61fb56122661fbe3) that the stack of C++20/23 `<format>` work will be done _eventually_. sweat_smile

That's almost done. The large part left is implementing the missing chrono parts so I can add formatters for these parts. Unfortunately I feel the library support for `<format>` is lacking, so guess who's working on papers to rectify that ;-)

You probably have noted I've a strong interest in the `std` and `std.compat` module too.

Adding support to `from_chars` is still on my todo list, but it's still low on that list and other new C++26 features may get a higher priority. I work on libc++ as a volunteer so things that I care about will get a higher priority.

**Comment by h-vetinari** [comment]
> I work on libc++ as a volunteer so things that I care about will get a higher priority.

Completely understandable! :)

**Comment by h-vetinari** [comment]
I'm wondering if it wouldn't make sense to keep this issue open as a tracker for finishing the implementation of P0067R5, particularly the `from_chars` part? It already has a bunch of relevant context and there's no other issue for that[^1].

[^1]: Yes, I know that there's a status page, but that page does not contain useful references like [`87c0160`](https://github.com/llvm/llvm-project/commit/87c016078ad72c46505461e4ff8bfa04819fe7ba), for example.

**Comment by mordante** [comment]
I prefer to keep it closed, we don't use the issue tracker to track libc++'s implementation status. We have the status page for that. Having some missing features as a bugs will cause us to forget closing bugs.
